### PR TITLE
Checkpoints in Sync

### PR DESCRIPTION
**Pull request by JimSuplizio**
### What does this PR do?
There are several things here
1) Allow the user to add/delete/modify checkpoints locally to their heart's content. When they're ready they can execute an SFDX: Create Checkpoints or hit the cloud-upload icon on the checkpoints. Both are disabled if checkpoints are not enabled for the project.

2) Source is not loaded until the very last moment. For checkpoints source/line information is loaded every time the user uploads the checkpoints via the SFDX command or cloud-upload icon. For replay debugging this is when the starts debugging. For replay debugging, this is how things worked prior to one of my previous changes.

3) Added a batch delete command to batch delete checkpoints. One call to get any existing checkpoints and one call to nuke them all. Note: For the creation of the new checkpoints I'm still calling one at a time, error processing is less of a pain than it is batching them for the creates.

### What issues does this PR fix or reference?
@W-4919221, W-4949697, W-4858717@

### CODE REVIEW

File: packages/salesforcedx-vscode-replay-debugger/src/messages/i18n.ts [hunk 1]
```diff
@@ -25,11 +25,14 @@ export const messages = {
     'The log file either is missing or does not have any log lines in it.',
   incorrect_log_levels_text:
     'The log must be generated with log categories Apex code at the FINEST level and Visualforce at the FINER level.',
-  up_to_five_checkpoints: 'You can set up to five checkpoints.',
+  up_to_five_checkpoints:
+    'You have %s of the maximum 5 enabled checkpoints. Remove or disable checkpoints until 5 or fewer are active.',
   checkpoints_can_only_be_on_valid_apex_source:
-    'Checkpoints can be set only on a valid line of Apex source.',
+    'Checkpoints may be set on a valid line of Apex source. URI=%s, Line=%s',
```
Comment (ruthemmanuelle) [inline packages/salesforcedx-vscode-replay-debugger/src/messages/i18n.ts]: Hi @JimSuplizio . We avoid "may" because it indicates permission. What is the reason for this change? Also, maybe we should give some context for why we're providing the values at the end of the message?

File: packages/salesforcedx-vscode-replay-debugger/src/messages/i18n.ts [hunk 2]
```diff
@@ -25,11 +25,14 @@ export const messages = {
...
   local_source_is_out_of_sync_with_the_server:
     "The local source is out of sync with the server. Push any changes you've made locally to your org, and pull any changes you've made in the org into your local project.",
+  checkpoints_successfully_uploaded:
+    'All checkpoints have been successfully uploaded to the server',
```
Comment (ruthemmanuelle) [inline packages/salesforcedx-vscode-replay-debugger/src/messages/i18n.ts]: Can we end this sentence in a period?

File: packages/salesforcedx-vscode-replay-debugger/src/messages/i18n.ts (hunk 1 above)
Comment (JimSuplizio) [reply packages/salesforcedx-vscode-replay-debugger/src/messages/i18n.ts]: I'll change it.

File: packages/salesforcedx-vscode-replay-debugger/src/messages/i18n.ts (hunk 2 above)
Comment (JimSuplizio) [reply packages/salesforcedx-vscode-replay-debugger/src/messages/i18n.ts]: Yup. I'll change it.

File: packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts [hunk 3]
```diff
@@ -132,76 +138,66 @@ export class CheckpointService implements TreeDataProvider<BaseNode> {
...
-  public canAddCheckpointNote(): Boolean {
-    return this.getChildren().length < MAX_ALLOWED_CHECKPOINTS;
+  public hasFiveOrLessActiveCheckpoints(displayError: boolean): Boolean {
+    let numEnabledCheckpoints = 0;
+    for (const cpNode of this.getChildren() as CheckpointNode[]) {
+      if (cpNode.isCheckpointEnabled()) {
+        numEnabledCheckpoints++;
+      }
+    }
+    const fiveOrLess: boolean =
```
Comment (DatGuyJonathan) [inline packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts]: Boolean type is implicit; you don't need to declare it.

File: packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts [hunk 4]
```diff
@@ -226,71 +222,34 @@ export class CheckpointService implements TreeDataProvider<BaseNode> {
...
+            nls.localize('local_source_is_out_of_sync_with_the_server')
           );
-        // This is the error code we'll see when the local source and the server are out of sync
-        // This means that the local type does not exist on the server. Give the error a more useful
-        // message as to why they can't set the checkpoint.
-        } else if (result[0].errorCode === FIELD_INTEGRITY_EXCEPTION) {
-          theNode.setActionCommandResultFailure(nls.localize('local_source_is_out_of_sync_with_the_server'));
         } else {
-          theNode.setActionCommandResultFailure(result[0].message);
+          vscode.window.showErrorMessage(
```
Comment (DatGuyJonathan) [inline packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts]: You can use template strings: `${result[0].message}. URI=${theNode.getCheckpointUri()}, Line=${theNode.getCheckpointLineNumber()}`

File: packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts [hunk 5]
```diff
@@ -322,26 +281,64 @@ export class CheckpointService implements TreeDataProvider<BaseNode> {
...
-                undefined,
-                record.Id
+            // If there are things to delete then create the batchRequest
+            if (successResult.records.length > 0) {
+              const requests: BatchRequest[] = [];
+              for (const record of successResult.records) {
+                const request: BatchRequest = {
+                  method: RestHttpMethodEnum.Delete,
+                  url:
+                    'services/data/v43.0/tooling/sobjects/ApexExecutionOverlayAction/' +
```
Comment (DatGuyJonathan) [inline packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts]: I thought you had a constant for this?

File: packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts [hunk 6]
```diff
@@ -500,84 +567,196 @@ export class CheckpointInfoIterationNode extends CheckpointInfoNode {
...
+      }
+    } else {
+      // The breakpoint is not a checkpoint any longer, call to delete it from the checkpoint list if it exists
+      checkpointService.deleteCheckpointNodeIfExists(breakpointId);
+    }
+  }
+
+  for (const bp of breakpointsChangedEvent.added) {
+    if (bp.condition && bp.condition!.toLowerCase().indexOf(CHECKPOINT) >= 0) {
+      if (bp instanceof vscode.SourceBreakpoint) {
```
Comment (DatGuyJonathan) [inline packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts]: You can include this in the first if statement

File: packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts (hunk 3 above)
Comment (vazexqi) [reply packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts]: This should fail the `npm lint` as well because we turn on https://palantir.github.io/tslint/rules/no-inferrable-types/ on https://github.com/forcedotcom/salesforcedx-vscode/blob/develop/tslint.json#L22

File: packages/salesforcedx-vscode-replay-debugger/src/commands/batchDeleteExistingOverlayActionsCommand.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+export interface BatchRequests {
+  batchRequests: BatchRequest[];
+}
+
+export interface BatchRequest {
+  method: RestHttpMethodEnum;
+  url: string;
+}
+
+// The batch response is a bit nuts. There's an array of results which contain an status code and a result however
```
Comment (DatGuyJonathan) [inline packages/salesforcedx-vscode-replay-debugger/src/commands/batchDeleteExistingOverlayActionsCommand.ts]: Remove snarky comments. This is an open source project and you’re representing Salesforce.

File: packages/salesforcedx-vscode-replay-debugger/src/index.ts [hunk 8]
```diff
@@ -46,37 +47,44 @@ function registerDebugHandlers(): vscode.Disposable {
...
 
-export async function activate(context: vscode.ExtensionContext) {
-  console.log('Apex Replay Debugger Extension Activated');
-  const commands = registerCommands();
-  const debugHandlers = registerDebugHandlers();
-  const debugConfigProvider = vscode.debug.registerDebugConfigurationProvider(
-    'apex-replay',
-    new DebugConfigurationProvider()
+  const config = vscode.workspace.getConfiguration();
+  const checkpointsEnabled = config.get(
```
Comment (DatGuyJonathan) [inline packages/salesforcedx-vscode-replay-debugger/src/index.ts]: You already do this in `activate()`, just pass it in to this method instead of doing it again.

File: packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts (hunk 6 above)
Comment (JimSuplizio) [reply packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts]: Changed that and the a similar if statement for the changed breakpoints

File: packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts (hunk 5 above)
Comment (JimSuplizio) [reply packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts]: You're right, I have OVERLAY_ACTION_DELETE_URL. Done.

File: packages/salesforcedx-vscode-replay-debugger/src/commands/batchDeleteExistingOverlayActionsCommand.ts (hunk 7 above)
Comment (JimSuplizio) [reply packages/salesforcedx-vscode-replay-debugger/src/commands/batchDeleteExistingOverlayActionsCommand.ts]: done

File: packages/salesforcedx-vscode-replay-debugger/src/index.ts [hunk 9]
```diff
@@ -127,7 +140,10 @@ function getDialogStartingPath(): vscode.Uri | undefined {
   }
 }
 
-async function retrieveLineBreakpointInfo(): Promise<boolean> {
+// Call to retrieve line/breakpoint information when the user decides to upload
+// their checkpoints. This ends up being done every time so we don't end up having
+// to perform any shenanigans when the user changes their code.
```
[the remark this answers is not in the document]
Comment (JimSuplizio) [reply packages/salesforcedx-vscode-replay-debugger/src/index.ts]: done

File: packages/salesforcedx-vscode-replay-debugger/src/index.ts (hunk 8 above)
Comment (JimSuplizio) [reply packages/salesforcedx-vscode-replay-debugger/src/index.ts]: done

File: packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts (hunk 4 above)
Comment (JimSuplizio) [reply packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts]: I'll make the change but here's one reason I don't like template strings. Say you're changing a getCheckpointUri to a different name, Find All References will not find the references that are in the template strings.

Comment (DatGuyJonathan) [reply packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts]: It works for me.

image

What do you see?

Comment (JimSuplizio) [reply packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts]: The wrong single quotes on mine. You are correct.

File: packages/salesforcedx-vscode-replay-debugger/src/constants.ts [hunk 10]
```diff
@@ -25,7 +24,10 @@ export const EVENT_VARIABLE_SCOPE_BEGIN = 'VARIABLE_SCOPE_BEGIN';
 export const EVENT_VARIABLE_ASSIGNMENT = 'VARIABLE_ASSIGNMENT';
 export const EXEC_ANON_SIGNATURE = 'execute_anonymous_apex';
 export const FIELD_INTEGRITY_EXCEPTION = 'FIELD_INTEGRITY_EXCEPTION';
+export const INVALID_CROSS_REFERENCE_KEY = 'INVALID_CROSS_REFERENCE_KEY';
 export const GET_LINE_BREAKPOINT_INFO_EVENT = 'getLineBreakpointInfo';
 export const LINE_BREAKPOINT_INFO_REQUEST = 'lineBreakpointInfo';
+export const OVERLAY_ACTION_DELETE_URL =
+  'services/data/v43.0/tooling/sobjects/ApexExecutionOverlayAction/';
```
Comment (DatGuyJonathan) [inline packages/salesforcedx-vscode-replay-debugger/src/constants.ts]: It would be nice if this was used in the tests as well.

File: packages/salesforcedx-vscode-replay-debugger/test/unit/checkpoints/checkpointService.test.ts [hunk 11]
```diff
@@ -5,8 +5,8 @@
  * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
 
-import { expect } from 'chai';
-import * as sinon from 'sinon';
+import { assert, expect } from 'chai';
+//import * as sinon from 'sinon';
```
Comment (DatGuyJonathan) [inline packages/salesforcedx-vscode-replay-debugger/test/unit/checkpoints/checkpointService.test.ts]: Can this be removed?

File: packages/salesforcedx-vscode-replay-debugger/src/messages/i18n.ts [hunk 12]
```diff
@@ -25,11 +25,14 @@ export const messages = {
     'The log file either is missing or does not have any log lines in it.',
   incorrect_log_levels_text:
     'The log must be generated with log categories Apex code at the FINEST level and Visualforce at the FINER level.',
-  up_to_five_checkpoints: 'You can set up to five checkpoints.',
+  up_to_five_checkpoints:
+    'You have %s of the maximum 5 enabled checkpoints. Remove or disable checkpoints until 5 or fewer are active.',
   checkpoints_can_only_be_on_valid_apex_source:
-    'Checkpoints can be set only on a valid line of Apex source.',
+    'Checkpoints can be set on a valid line of Apex source. URI=%s, Line=%s',
```
Comment (ruthemmanuelle) [inline packages/salesforcedx-vscode-replay-debugger/src/messages/i18n.ts]: Thanks for changing "may" to "can", @JimSuplizio . I'm still not sure the user will understand why we're including those values at the end. Can we prefix them with a label? Maybe something like this (if it's accurate)?
`'Checkpoints can be set only on a valid line of Apex source. Invalid checkpoint location: URI=%s, Line=%s'`

File: packages/salesforcedx-vscode-replay-debugger/src/constants.ts (hunk 10 above)
Comment (JimSuplizio) [reply packages/salesforcedx-vscode-replay-debugger/src/constants.ts]: Thanks for kindly pointing that out to me. Done.

File: packages/salesforcedx-vscode-replay-debugger/test/unit/checkpoints/checkpointService.test.ts (hunk 11 above)
Comment (JimSuplizio) [reply packages/salesforcedx-vscode-replay-debugger/test/unit/checkpoints/checkpointService.test.ts]: done

File: packages/salesforcedx-vscode-replay-debugger/src/messages/i18n.ts (hunk 12 above)
Comment (JimSuplizio) [reply packages/salesforcedx-vscode-replay-debugger/src/messages/i18n.ts]: done.

**Comment by vazexqi** [comment]
I think there are missing copyright notices in the new files that were added. I randomly picked one of the new test files and I didn't see a copyright header on it.

Could you fix it in a subsequent commit?
